src/hotspot/share/runtime/thread.cpp

```diff
@@ -1345,6 +1336,10 @@ void NonJavaThread::pre_run() {
 void NonJavaThread::post_run() {
   JFR_ONLY(Jfr::on_thread_exit(this);)
   remove_from_the_list();
+  unregister_thread_stack_with_NMT();
+  set_stack_base(NULL);
+  set_stack_size(0);
```

Comment:
Sorry but I just don't see any need for clearing stack_base or stack_size, and then you don't need to mess with the assertion in stack_base().

Reply:
I agree with David that the calls to `set_stack_base(NULL); set_stack_size(0);` do not seem to be related to this bug. Also, previously only `set_stack_base(NULL);` was called -- on all Threads. Now:

- NonJavaThread: `set_stack_base(NULL); set_stack_size(0);`
- JavaThread: neither operation is done

I am wondering:

- why the behavior is changed
- is zeroing the stack_base/size necessary for NMT code only, or is it necessary for non-NMT operations as well?

Will any test cases fail if you remove line 1340/1341?

Reply:
I've tried without clearing stack_base and stack_size and also reverted the change in stack_base() in thread.hpp.
It passed tier1 testing. I'll run more test.

Reply:
> Sorry but I just don't see any need for clearing stack_base or stack_size, and then you don't need to mess with the assertion in stack_base().

Well, without cleaning stack_base and stack_size, the thread appears to be live and running, which contradicts NMT report (and maybe others, e.g. smap, etc.)

My initial thought is to filter out terminated thread in Threads::print_on_error(), but it does not seem to have a way.

Reply:
- JavaThread guarantees to delete 'this' before thread exits.
- NonJavaThread object can outlive thread. Zeroing thread stack/size is a way to indicate liveness of the actual thread. I was chasing a race before I realized the thread actual dead, very confusing to see 'G1 Main Marker' as live thread in hs_err file.